### PR TITLE
Introduce `RzArchPlatform` and support for platform profiles

### PR DESCRIPTION
**Pull request by officialcjunior**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

Introduces `RzArchPlatformTarget` and `RzArchPlatformItem` for platform profiles.

The platform profiles stored in `librz/asm/platforms` following a file naming convention `arch-cpu-platform` are firstly loaded up at `rz_arch_platform_init`. Subsequently, it's parsed and stored in a hashtable inside `RzArchPlatformTarget` which is a pair of `ut64` address and a `RzArchPlatformItem`. `RzArchPlatformItem` is a struct/module which houses the name and the comment(if exists) of the corresponding port. The names are added as flags and comments as comments(`CCu`) in `rz_arch_platform_add_flags_comments`.

A new configuration variable `asm.platform` is realized to choose the platform profile : fdf3c0286b96057b96572a39ed4efd6cdb26dedf. 
To support the options to choose from, a new variable `platforms` was added to `RzAsmPlugin` (similar to `features` and `cpus`). This is where the name of the platforms supported by each of the architectures/`RzAsmPlugin` will be defined (also similar to how `cpus` work).

Currently, Broadcom's [BCM2835 ARM Peripherals](https://www.raspberrypi.org/documentation/hardware/raspberrypi/bcm2835/README.md), the one that runs in most of the Raspberry Pi-s, is the only profile that is added. More can be added when the code review is finished.

**Test plan**

- [x] Add more platforms.

**Closing issues**

None

### CODE REVIEW

File: librz/analysis/arch_platform.c [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+
+/**
+ * \brief Frees an RzArchPlatformItem type
+ */
+RZ_API void rz_arch_platform_item_free(RzArchPlatformItem *item) {
+	if (!item) {
+		return;
+	}
+	free(item->name);
+	free(item->comment);
```
Comment (XVilka) [inline librz/analysis/arch_platform.c:54]: Should it free also the `item` itself?

File: librz/analysis/arch_platform.c [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+				item->comment = comment;
+			}
+			ht_up_insert(t->platforms, address, item);
+		}
+	}
+	return true;
+}	
+
+static bool sdb_load_arch_platform_by_path(RZ_NONNULL RzArchPlatformTarget *t, const char *path) {
+	Sdb *db = sdb_new(0, path, 0);
```
Comment (XVilka) [inline librz/analysis/arch_platform.c:97]: Please also check if `db` is NULL.

Comment (XVilka) [reply librz/analysis/arch_platform.c:97]: I also recommend to check for `path` not being NULL and add `RZ_NONNULL` too.

File: librz/analysis/arch_platform.c [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+}
+
+/**
+ * \brief Loads the contents of the Platform Profile to the RzArchPlatformTarget
+ *
+ * \param t reference to RzArchPlatformTarget
+ * \param path reference to path of the SDB file
+ */
+RZ_API bool rz_arch_load_platform_sdb(RzArchPlatformTarget *t, const char *path) {
+	if (!rz_file_exists(path)) {
```
Comment (XVilka) [inline librz/analysis/arch_platform.c:118]: Same here.

File: librz/analysis/arch_platform.c [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+
+/**
+ * \brief Initialize Platform Profiles by setting the path to the corresponding SDB file
+ * 
+ * \param t reference to RzArchPlatformTarget
+ * \param arch reference to the selected architecture (value of `asm.arch`
+ * \param platform reference to the selected platform (value of `asm.platform`)
+ * \param dir_prefix reference to the directory prefix or the value of dir.prefix
+ */
+RZ_API bool rz_arch_platform_init(RzArchPlatformTarget *t, const char *arch, const char *cpu, const char *platform, const char *dir_prefix) {
```
Comment (XVilka) [inline librz/analysis/arch_platform.c]: Add `rz_return_val_if_fail()` with arch, cpu, platform, dir_prefix.

File: librz/core/canalysis.c [hunk 5]
```diff
@@ -6479,6 +6479,29 @@ RZ_API void rz_arch_profile_add_flag_every_io(RzArchProfile *profile, RzFlag *fl
 	ht_up_foreach(profile->registers_extended, add_mmio_extended_flag_cb, flags);
 }
 
+static bool add_arch_platform_flag_comment_cb(void *user, const ut64 addr, const void *v) {
```
Comment (XVilka) [inline librz/core/canalysis.c:6482]: Please check for `v` not being NULL since you dereference it later in the code.

File: librz/core/cconfig.c [hunk 6]
```diff
@@ -762,6 +769,45 @@ static bool cb_asmfeatures(void *user, void *data) {
...
+					rz_list_append(node->options, strdup(feature));
+				}
+			}
+			free(platforms);
+		}
+	}
+}
+
+static bool cb_asmplatform(void *user, void *data) {
+	RzCore *core = (RzCore *)user;
```
Comment (XVilka) [inline librz/core/cconfig.c:793]: Same here, check for them not being NULL.

File: librz/include/rz_arch.h [hunk 7]
```diff
@@ -41,6 +49,13 @@ RZ_API bool rz_arch_profiles_init(RzArchTarget *c, const char *cpu, const char *
 RZ_API void rz_arch_profile_add_flag_every_io(RzArchProfile *profile, RzFlag *flags);
 RZ_API bool rz_arch_load_profile_sdb(RzArchTarget *t, const char *path);
 
+RZ_API RZ_OWN RzArchPlatformItem *rz_arch_platform_item_new(RZ_NULLABLE const char *name);
+RZ_API RZ_OWN RzArchPlatformTarget *rz_arch_platform_target_new();
+RZ_API void rz_arch_platform_target_free(RzArchPlatformTarget *target);
+RZ_API void rz_arch_platform_item_free(RzArchPlatformItem *item);
+RZ_API bool rz_arch_load_platform_sdb(RzArchPlatformTarget *t, const char *path);
+RZ_API bool rz_arch_platform_init(RzArchPlatformTarget *t, const char *arch, const char *cpu, const char *platform, const char *dir_prefix);
```
Comment (XVilka) [inline librz/include/rz_arch.h]: Also please add these `RZ_NONNULL` here as well.

**Review by XVilka — CHANGES_REQUESTED** [review]

File: librz/analysis/arch_platform.c [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+				item->comment = comment;
+			}
+			ht_up_insert(t->platforms, address, item);
+		}
+	}
+	return true;
+}
+
+static bool sdb_load_arch_platform_by_path(RZ_NONNULL RzArchPlatformTarget *t, RZ_NONNULL const char *path) {
+	if (!path) {
```
Comment (XVilka) [inline librz/analysis/arch_platform.c:94]: Maybe add here also `rz_return_val_if_fail()`

File: librz/analysis/arch_platform.c [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+}
+
+/**
+ * \brief Loads the contents of the Platform Profile to the RzArchPlatformTarget
+ *
+ * \param t reference to RzArchPlatformTarget
+ * \param path reference to path of the SDB file
+ */
+RZ_API bool rz_arch_load_platform_sdb(RzArchPlatformTarget *t, RZ_NONNULL const char *path) {
+	if (!path) {
```
Comment (XVilka) [inline librz/analysis/arch_platform.c:115]: Same here.

File: librz/analysis/arch_platform.c [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+
+/**
+ * \brief Initialize Platform Profiles by setting the path to the corresponding SDB file
+ * 
+ * \param t reference to RzArchPlatformTarget
+ * \param arch reference to the selected architecture (value of `asm.arch`
+ * \param platform reference to the selected platform (value of `asm.platform`)
+ * \param dir_prefix reference to the directory prefix or the value of dir.prefix
+ */
+RZ_API bool rz_arch_platform_init(RzArchPlatformTarget *t, const char *arch, const char *cpu, const char *platform, const char *dir_prefix) {
```
Comment (XVilka) [inline librz/analysis/arch_platform.c]: Mark them please as `RZ_NONNULL`

**Review by XVilka — APPROVED** [review]
Overall is good, please address a couple more nitpicks.

**Comment by XVilka** [comment]
Please also craft a few tests for that.

File: librz/include/rz_arch.h (hunk 7 above)
Comment (officialcjunior) [reply librz/include/rz_arch.h]: Where exactly should I add `RZ_NONNULL` here?

File: librz/analysis/arch_platform.c (hunk 10 above)
Comment (officialcjunior) [reply librz/analysis/arch_platform.c]: For all of the function's arguments?

File: librz/include/rz_arch.h (hunk 7 above)
Comment (XVilka) [reply librz/include/rz_arch.h]: To `arch`, `cpu`, etc

File: librz/analysis/arch_platform.c (hunk 10 above)
Comment (XVilka) [reply librz/analysis/arch_platform.c]: Not for all, for `arch`, `cpu`, etc

File: librz/include/rz_arch.h (hunk 7 above)
Comment (wargio) [reply librz/include/rz_arch.h]: RZ_NONNULL goes to any value that you check when you use `rz_return[_val]_if()`

**Comment by officialcjunior** [comment]
The reason why one of the ASAN tests fail is probably because the platform is NULL until it's set:
Here's a demonstration:
```
$ ./build/binrz/rizin/rizin test/bins/arm/elf/hello_world
(..)
/usr/local/share/rizin/0.3.0-git/asm/platforms/x86-x86-.sdb
(..)
/usr/local/share/rizin/0.3.0-git/asm/platforms/arm-v8-.sdb
 -- Execute commands on a temporary offset by appending '@ offset' to your command.
[0x000003fc]> e asm.cpu=arm1176
/usr/local/share/rizin/0.3.0-git/asm/platforms/arm-arm1176-.sdb
[0x000003fc]> e asm.platform=bcm2835
/usr/local/share/rizin/0.3.0-git/asm/platforms/arm-arm1176-bcm2835.sdb
```
The CPU (v8) is automatically set when the arch is found to be arm, but not the platform.

Any idea on how to fix that?

**Comment by XVilka** [comment]
@officialcjunior it's an architectural problem of Rizin, imho. Meanwhile, we could just remove `rz_return_if_fail()` for `platform` and silently exit if it's NULL.

**Review by XVilka — APPROVED** [review]

**Review by wargio — APPROVED** [review]

**Pull request merged**